Userland/keymap.cpp

```diff
@@ -41,12 +41,24 @@ char* read_map(const JsonObject& json, const String& name)
     return map;
 }
 
-int read_map_from_file(String filename)
+int read_map_from_file_or_name(String filename)
 {
     auto file = CFile::construct(filename);
     if (!file->open(CIODevice::ReadOnly)) {
-        fprintf(stderr, "Failed to open %s: %s\n", filename.characters(), file->error_string());
-        return 1;
+        if (!filename.ends_with(".json")) {
+            StringBuilder full_path;
+            full_path.append("/res/keymaps/");
+            full_path.append(filename);
+            full_path.append(".json");
+            filename = full_path.to_string();
+            file = CFile::construct(filename);
+            if (!file->open(CIODevice::ReadOnly))
+                goto error;
+        } else {
+error:
```

Comment:
This feels very muddy. Please restructure the program so you don't need to `goto` like this :)

For example, you could make a helper function that does the filename fallback:
```cpp
RefPtr<CFile> open_keymap_file(const String& name)
{
    auto file = CFile::construct(filename);
    if (file->open(CIODevice::ReadOnly))
        return file;
    // try the automatic path expansion etc...
}

int read_map_from_file(const String& name)
{
    auto file = open_keymap_file(name);
    if (!file) {
        fprintf(stderr, "error message\n");
        return 1;
    }
    // do stuff with file
}

Reply:
Thanks for the proposal! It worked out perfectly minus `const String&` arg type (left as `String`) and `if (!file)` (changed to `if (!file->is_open())`); see 78a54b4

Reply:
Seems `String&` by itself works fine as to show `New keymap loaded from "/res/keymap/fi.json".` for example instead of `New keymap loaded from "fi".` (just given input)
I'll push the changes soon ^^